modules/commons/src/main/java/org/apache/synapse/commons/vfs/VFSUtils.java

```diff
@@ -483,6 +483,121 @@ public static FileSystemOptions attachFileSystemOptions(Map<String, String> opti
         return opts;
     }
 
+    /**
+     * Function to resolve hostname of the vfs uri
+     * @param uri
+     * @return
+     * @throws FileSystemException
+     * @throws UnknownHostException
+     */
+    public static String resolveUriHost (String uri) throws FileSystemException, UnknownHostException {
+        return resolveUriHost(uri, new StringBuilder());
+    }
+
+    /**
+     * Function to resolve the hostname of uri to ip for following vfs protocols. if not the protocol listed, return
+     * same uri provided for {uri}
+     * Protocols resolved : SMB
+     * @param uri URI need to resolve
+     * @param strBuilder
+     * @return hostname resolved uri
+     * @throws FileSystemException
+     * @throws UnknownHostException
+     */
+    public static String resolveUriHost (String uri, StringBuilder strBuilder)
+            throws FileSystemException, UnknownHostException {
+
+        if (uri != null && strBuilder != null) {
+            // Extract the scheme
+            String scheme = UriParser.extractScheme(uri, strBuilder);
+
+            //need to resolve hosts of smb URIs due to limitation in jcifs library
+            if (scheme != null && (scheme.equals("smb"))) {
+                // Expecting "//"
+                if (strBuilder.length() < 2 || strBuilder.charAt(0) != '/' || strBuilder.charAt(1) != '/') {
+                    throw new FileSystemException("vfs.provider/missing-double-slashes.error", uri);
+                }
+                strBuilder.delete(0, 2);
+
+                // Extract userinfo
+                String userInfo = extractUserInfo(strBuilder);
+
+                // Extract hostname
+                String hostName = extractHostName(strBuilder);
+
+                //resolve host name
+                InetAddress hostAddress = InetAddress.getByName(hostName);
+                String resolvedHostAddress = hostAddress.getHostAddress();
+
+                //build resolved uri
+                StringBuilder uriStrBuilder = new StringBuilder();
+                uriStrBuilder.append(scheme).append("://");
+
+                if (userInfo != null) {
+                    //user information can be null since it's optional
+                    uriStrBuilder.append(userInfo).append("@");
+                }
+
+                uriStrBuilder.append(resolvedHostAddress).append(strBuilder);
+
+                return uriStrBuilder.toString();
+            }
+        }
+
+        return uri;
+    }
+
+
+    /**
+     * Extracts the hostname from a URI.  The scheme://userinfo@ part has
+     * been removed.
+     * extracted hostname will be reoved from the StringBuilder
+     */
+    private static String extractHostName(StringBuilder name) {
+        final int maxlen = name.length();
+        int pos = 0;
+        for (; pos < maxlen; pos++) {
+            char ch = name.charAt(pos);
+            if (ch == '/' || ch == ';' || ch == '?' || ch == ':'
```

Comment:
Thanks for pointing, I stole the logic from wso2-commons-vfs :-D 
https://github.com/wso2/wso2-commons-vfs/blob/master/core/src/main/java/org/apache/commons/vfs2/provider/HostFileNameParser.java#L194

I'll describe the logic in comment as you suggest